neps/nep-0399.md

```diff
@@ -0,0 +1,340 @@
+  from FlatStorage can be combined with the improvement brought by these other ideas,
+  so the implementation of FlatStorage won't be rendered obsolete in the future.
+
+Q: What other designs have been considered and what is the rationale for not choosing them?
+
+A: Alternatively, we can still get rid of TTN cost by increasing the base fees for storage reads and writes. However, 
+  this could require increasing the fees by quite a lot, which could end up breaking many contracts.
+Q: What is the impact of not doing this?
+
+A: Storage reads will remain inefficiently implemented and cost more than it could be.
+
+## Specification
+The key idea of FlatStorage is to store a direct mapping from trie keys to values on disk. 
+Here the values of this mapping can be either the value corresponding to the trie key itself, 
+or the value ref, a hash that points to the address of the value. If the value itself is stored, 
+only one disk read is needed to look up a value from flat storage, otherwise two disk reads if the value 
+ref is stored. We will discuss more in the following section for whether we use values or value refs. 
+For the purpose of high level discussion, it suffices to say that with FlatStorage, 
+at most two disk reads are needed to perform a storage read.
+
+The simple design above won't work because there could be forks in the chain. In the following case, FlatStorage 
+must support key value lookups for states of the blocks on both forks.
+```
+           Block B1 - Block B2 - ...
+        / 
+block A 
+        \  Block C1 - Block C2 - ...
+```
+
+The handling of forks will be the main consideration of the following design. More specifically,
+the design should satisfy the following requirements,
+1) It should support concurrent block processing. Blocks on different forks are processed 
+   concurrently in our client code, so the flat storage API must support that.
+2) In case of long forks, block processing time should not be too much longer than the average case. 
+   We don’t want this case to be exploitable. It is acceptable that block processing time is 200ms longer, 
+   which may slow down block production, but probably won’t cause missing blocks and chunks. 
+   It is not acceptable if block processing time is 10s, which may lead to more forks and instability in the network.
+3) The design must be able to decrease storage access cost in all cases, 
+   since we are going to change the storage read fees based on flat storage. 
+   We can't conditionally enable FlatStorage for some blocks and disable it for other, because 
+   the fees we charge must be consistent. 
+   
+The mapping of key value pairs FlatStorage stored on disk matches the state at some block. 
+We call this block the head of flat storage, or the flat head. There are two natural options 
+for which block should be the flat head, the chain head, or the last final block. Although
```

Comment:
Should we explain the terms chain head and last final block somewhere?  Alternatively, I propose that we do not discuss chain head at all here.